api.md

```diff
@@ -0,0 +1,185 @@
+
+## Response
+
+- Status 200
+  - Header `Content-Type: application/cbor`
+
+# CCF based design
+
+## Operations table
+
+This is a singleton table, without keys. Each record can be one of the following:
+
+- `{ "status": "registered" }`: A claim was registered synchronously. The Operation ID and Entry ID are both equal to the transaction ID of this record.
+- `{ "status": "pending", "startTime": TIME, "contextDigest": "..." }`: A claim was submitted and is being processed asynchronously. The Operation ID is equal to the transaction ID of this record. The entry ID is not yet know.
+- `{ "operationId": "<Operation ID>", "status": "registered" }`: An asynchronous claim registration has completed successfully. The Entry ID is equal to the transaction ID of this record.
+- `{ "operationId": "<Operation ID>", "status": "failed", "error": <Error> }`: An asynchronous claim registration has failed.
+
+The fact that the Operation IDs and Entry IDs share the same domain and may
+overlap (as is the case for synchronous submissions) is an implementation
+detail of the CCF-based ledger. Other implementations may differ.
+
+## Operations indexing strategy
+
+In order to implement the `/operations/<Operation ID>` endpoint, an indexing
+strategy is implemented over the operations table. It constructs an in-memory
+map, from operation ID to status.
+
+The indexing strategy periodically purges old entries (either by count, or by
+time). It should keep track of the lower and upper bound of the transaction IDs
+it covers.
+
+If looking up an operation gives no result, than this can be one of three cases:
+- The ID is greater than the upper bound, meaning the operation has not yet been indexed. Returns a "pending" status.
+- The ID is lower then the lower bound, meaning the operation has been flushed. Return a "not found" status.
+- The ID falls in between the lower and upper bound, it must be an invalid operation ID. Return either status.
+
+TODO: can we detect view changes and leader elections?
+
+## Synchronous submission flow (X509 or cached DID)
+
+1) Client makes a `POST /entries` request.
+2) The ledger validates the claim and makes the following writes:
+    - Writes the claim to the `entry` table.
+    - Writes `{ "status": "registered" }` to the `operations` table.
+    - Sets the claims digest of the transaction to the receipt TBS.
+3) The local commit handler takes the transaction number and returns it to the
+    client in a JSON document as the operation ID and `"status": "pending"`.
+4) Client requests `GET /operations/<Operation ID>`
+    - The indexing strategy is unaware of the operation ID
+    - The endpoint returns `"status": "pending"`
+5) CCF replicates the transaction and its signature.
+6) The indexing strategy witnesses the write to the `operations` table and
+   updates its map, using the transaction ID as both the Operation and Entry ID.
+7) Client requests `GET /operations/<Operation ID>`. Using the indexing
```

Comment:
I'm wondering if there may still be an issue with how global commits work: https://microsoft.github.io/CCF/main/build_apps/kv/kv_how_to.html#global-commit. A global commit may be observed by a given node when then returns `registered` but a follow-up GET to another node that's a little behind may not have the current state yet.